Documentation/architecture.rst

```diff
@@ -0,0 +1,155 @@
+  provides a set of prefilter rules used to filter traffic from the network. Specifically,
+  a set of CIDR maps supplied by the Cilium agent are used to do a lookup and the packet
+  is either dropped or allowed to be processed by the stack. This can be easily
+  extended as needed to build in new prefilter criteria/capabilities.
+
+* **Endpoint Policy:** The endpoint policy object implements the Cilium endpoint enforcement.
+  Using a map to lookup a packets associated identity and policy this layer
+  scales well to lots of endpoints. Depending on the policy this layer may drop the
+  packet, forward to a local endpoint, forward to the service object or forward to the
+  L7 Policy object for further L7 rules. This is the primary object in the Cilium
+  datapath responsible for mapping packets to identities and enforcing L3 and L4 policies.
+
+* **Service:** The Service object performs a map lookup on the destination IP
+  and optionally destination port for every packet received by the object.
+  If a matching entry is found, the packet will be forwarded to one of the
+  configured L3/L4 endpoints. The Service block can be used to implement a
+  standalone load balancer on any interface using the TC ingress hook or may
+  be integrated in the endpoint policy object.
+
+* **Socked Layer Enforcement:** Socket layer enforcement use two
+  hooks the socket operations hook and the socket send/recv hook to monitor
+  and attach to all TCP sockets associated with Cilium managed endpoints, including
+  any L7 proxies. The socket operations hook
+  will identify candidate sockets for accelerating. These include all local node connections
+  (endpoint to endpoint) and any connection to a Cilium proxy.
+  These identified connections will then have all messages handled by the socket
+  send/recv hook and will be accelerated using sockmap fast redirects. The fast
+  redirect ensures all policies implemented in Cilium are valid for the associated
+  socket/endpoint mapping and assuming they are sends the message directly to the
+  peer socket. This is allowed because the sockmap send/recv hooks ensures the message
+  will not need to be processed by any of the objects above.
+
+* **L7 Policy:** The L7 Policy object redirect proxy traffic to a Cilium userspace
+  proxy instance. Cilium uses an Envoy instance as its userspace proxy. Envoy will
+  then either forward the traffic or drop the traffic based on the configured L7 policy.
+
+These components are connected to create the flexible and efficient datapath used
+by Cilium. Below we show the following possible flows connecting endpoints on a single
+node, ingress to an endpoint, and endpoint to egress networking device. In each case
+there is an additional diagram showing the TCP accelerated path available when Socket
+Layer Enforcement is enabled.
+
+First we show the local endpoint to endpoint flow with optional L7 Policy on
```

Comment:
These sections could probably be prefaced by subsections something like:

```
Endpoint to Endpoint
--------------------
```
...
```
Egress from Endpoint
--------------------
```
...
```
Ingress to Endpoint
-------------------
```

(Not sure if `----` or `~~~~` is the right level of subsectioning, I think it's `---`.)